From 469163c598d36259101c0caadc3fb0dcb2007b9a Mon Sep 17 00:00:00 2001
From: Fabio Erculiani <lxnay@sabayon.org>
Date: Wed, 26 Jun 2013 09:03:54 +0200
Subject: [PATCH] Clean up closing empty sessions when not running under
 systemd

In logind's manager_gc(), submit "closing" session to garbage collection, to
clean them up once they are empty. This is usually triggered through
manager_cgroup_notify_empty() from a org.freedesktop.systemd1.Agent.Released
signal, but that doesn't happen when not running under systemd.
---
 src/login/logind.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/login/logind.c b/src/login/logind.c
index 5a39440..db7a724 100644
--- a/src/login/logind.c
+++ b/src/login/logind.c
@@ -1371,9 +1371,17 @@ void manager_gc(Manager *m, bool drop_not_started) {
         Seat *seat;
         Session *session;
         User *user;
+        Iterator i;
 
         assert(m);
 
+        /* clean up empty sessions when not running under systemd */
+        if (sd_booted() <= 0) {
+                HASHMAP_FOREACH(session, m->session_cgroups, i)
+                        if (session_get_state(session) == SESSION_CLOSING)
+                                session_add_to_gc_queue(session);
+        }
+
         while ((seat = m->seat_gc_queue)) {
                 LIST_REMOVE(Seat, gc_queue, m->seat_gc_queue, seat);
                 seat->in_gc_queue = false;
-- 
1.8.2.1

